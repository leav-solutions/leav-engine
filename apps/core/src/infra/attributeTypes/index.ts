// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export {default as attributeAdvancedLink} from './attributeAdvancedLinkRepo';
export {default as attributeAdvanced} from './attributeAdvancedRepo';
export {default as attributeSimpleLink} from './attributeSimpleLinkRepo';
export {default as attributeSimple} from './attributeSimpleRepo';
export {default as attributeTree} from './attributeTreeRepo';
export {default} from './attributeTypesRepo';
