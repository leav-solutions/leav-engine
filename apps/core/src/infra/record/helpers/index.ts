// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export {default as filterTypes} from './filterTypes';
export {default as getClassifyingFiltersVariableQueryPart} from './getClassifyingFiltersVariableQueryPart';
export {default as getSearchVariableName} from './getSearchVariableName';
export {default as getSearchVariablesQueryPart} from './getSearchVariablesQueryPart';
