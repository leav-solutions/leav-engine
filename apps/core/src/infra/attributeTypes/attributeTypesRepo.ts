// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {AqlLiteral, AqlQuery, GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';
import {AttributeTypes, IAttribute} from '../../_types/attribute';
import {AttributeCondition, IRecordFilterOption, IRecordSort} from '../../_types/record';
import {IValue, IValuesOptions} from '../../_types/value';

// To avoid some cyclic dependencies issues, we have to pass repo along attribute props
export interface IAttributeWithRepo extends IAttribute {
    _repo: IAttributeTypeRepo;
}

export interface IAttributeTypesRepo {
    getTypeRepo?(attribute: IAttribute): IAttributeTypeRepo;
}

export type GetConditionPartParentFunc = (
    condition: AttributeCondition,
    value: string | number | boolean,
    attribute: IAttribute
) => GetConditionPartFunc;

export type GetConditionPartFunc = (valueIdentifier: string | AqlLiteral) => GeneratedAqlQuery;
/**
 * Define interface used for all attribute type specific files
 */
export interface IAttributeTypeRepo {
    createValue({
        library,
        recordId,
        attribute,
        value,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        value: IValue;
        ctx: IQueryInfos;
    }): Promise<IValue>;

    /**
     * Update an existing value. Field "id" is expected on the value
     */
    updateValue({
        library,
        recordId,
        attribute,
        value,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        value: IValue;
        ctx: IQueryInfos;
    }): Promise<IValue>;

    /**
     * Delete an existing value. Field "id" is expected on the value
     */
    deleteValue({
        library,
        recordId,
        attribute,
        value,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        value: IValue;
        ctx: IQueryInfos;
    }): Promise<IValue>;

    /**
     * Get all values for given record and attribute
     *
     * @return Array<{}>    Return an empty array if no value found
     */
    getValues({
        library,
        recordId,
        attribute,
        forceGetAllValues,
        options,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        forceGetAllValues?: boolean;
        options?: IValuesOptions;
        ctx: IQueryInfos;
    }): Promise<IValue[]>;

    /**
     * Return a specific value based on its ID. Field "id" is expect on the value
     *
     * @return {}   Return null if no value found
     */
    getValueById({
        library,
        recordId,
        attribute,
        valueId,
        ctx
    }: {
        library: string;
        recordId: string;
        attribute: IAttribute;
        valueId: string;
        ctx: IQueryInfos;
    }): Promise<IValue>;

    /**
     * Return AQL query part to filter on this attribute. If will be concatenate with other filters and full query
     */
    filterQueryPart(attributes: IAttributeWithRepo[], filter: IRecordFilterOption, parentIdentifier?: string): AqlQuery;

    /**
     * Return AQL query part to sort on this attribute
     */
    sortQueryPart({attributes, order}: IRecordSort): AqlQuery;

    /**
     * Clear all values of given attribute. Can be used to cleanup values when an attribute is deleted for example.
     *
     * @param attribute
     * @return Promise<number> TRUE if operation succeed
     */
    clearAllValues({attribute, ctx}: {attribute: IAttribute; ctx: IQueryInfos}): Promise<boolean>;
}

export const ATTRIB_COLLECTION_NAME = 'core_attributes';
export const BASE_QUERY_IDENTIFIER = 'r';

interface IDeps {
    'core.infra.attributeTypes.attributeSimple'?: IAttributeTypeRepo;
    'core.infra.attributeTypes.attributeSimpleLink'?: IAttributeTypeRepo;
    'core.infra.attributeTypes.attributeAdvanced'?: IAttributeTypeRepo;
    'core.infra.attributeTypes.attributeAdvancedLink'?: IAttributeTypeRepo;
    'core.infra.attributeTypes.attributeTree'?: IAttributeTypeRepo;
}

export default function ({
    'core.infra.attributeTypes.attributeSimple': attributeSimpleRepo = null,
    'core.infra.attributeTypes.attributeSimpleLink': attributeSimpleLinkRepo = null,
    'core.infra.attributeTypes.attributeAdvanced': attributeAdvancedRepo = null,
    'core.infra.attributeTypes.attributeAdvancedLink': attributeAdvancedLinkRepo = null,
    'core.infra.attributeTypes.attributeTree': attributeTreeRepo = null
}: IDeps = {}): IAttributeTypesRepo {
    return {
        getTypeRepo(attribute) {
            let attrTypeRepo: IAttributeTypeRepo;
            switch (attribute.type) {
                case AttributeTypes.SIMPLE:
                    attrTypeRepo = attributeSimpleRepo;
                    break;
                case AttributeTypes.SIMPLE_LINK:
                    attrTypeRepo = attributeSimpleLinkRepo;
                    break;
                case AttributeTypes.ADVANCED:
                    attrTypeRepo = attributeAdvancedRepo;
                    break;
                case AttributeTypes.ADVANCED_LINK:
                    attrTypeRepo = attributeAdvancedLinkRepo;
                    break;
                case AttributeTypes.TREE:
                    attrTypeRepo = attributeTreeRepo;
                    break;
            }

            return attrTypeRepo;
        }
    };
}
