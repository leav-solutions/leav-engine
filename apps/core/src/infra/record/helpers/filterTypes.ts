// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {AttributeCondition, IRecordFilterOption, TreeCondition} from '../../../_types/record';

export interface IFilterTypesHelper {
    isAttributeFilter(filter: IRecordFilterOption): boolean;
    isClassifyingFilter(filter: IRecordFilterOption): boolean;
    isCountFilter(filter: IRecordFilterOption): boolean;
}

export default function(): IFilterTypesHelper {
    const _isAttributeFilter = (filter: IRecordFilterOption) => {
        return (filter as IRecordFilterOption).condition in AttributeCondition;
    };

    const _isClassifyingFilter = (filter: IRecordFilterOption) => {
        return (filter as IRecordFilterOption).condition in TreeCondition;
    };

    const _isCountFilter = (filter: IRecordFilterOption) => {
        return [
            AttributeCondition.VALUES_COUNT_EQUAL,
            AttributeCondition.VALUES_COUNT_GREATER_THAN,
            AttributeCondition.VALUES_COUNT_LOWER_THAN,
            AttributeCondition.IS_EMPTY,
            AttributeCondition.IS_NOT_EMPTY
        ].includes(filter.condition as AttributeCondition);
    };

    return {
        isAttributeFilter: _isAttributeFilter,
        isClassifyingFilter: _isClassifyingFilter,
        isCountFilter: _isCountFilter
    };
}
