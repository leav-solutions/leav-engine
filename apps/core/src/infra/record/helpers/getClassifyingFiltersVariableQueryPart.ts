// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {GeneratedAqlQuery, literal} from 'arangojs/aql';
import {IDbService} from 'infra/db/dbService';
import {IRecordFilterOption} from '_types/record';
import {getEdgesCollectionName, getFullNodeId, getRootId} from '../../../infra/tree/helpers/utils';
import {MAX_TREE_DEPTH} from '../../../infra/tree/treeRepo';
import {NODE_LIBRARY_ID_FIELD, NODE_RECORD_ID_FIELD} from '../../../infra/tree/_types';
import {IFilterTypesHelper} from './filterTypes';

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.record.helpers.filterTypes'?: IFilterTypesHelper;
}

export type GetClassifyingFiltersVariableQueryPart = (filter: IRecordFilterOption) => GeneratedAqlQuery;

export default function ({
    'core.infra.record.helpers.filterTypes': filterTypesHelper = null,
    'core.infra.db.dbService': dbService = null
}: IDeps): GetClassifyingFiltersVariableQueryPart {
    return filter => {
        if (!filterTypesHelper.isClassifyingFilter(filter)) {
            return null;
        }

        const collec = dbService.db.collection(getEdgesCollectionName(filter.treeId));

        const startingNode = filter.value
            ? getFullNodeId(String(filter.value), filter.treeId)
            : getRootId(filter.treeId);

        const queryPart = aql`
            FOR v, e IN 1..${MAX_TREE_DEPTH} OUTBOUND ${startingNode}
                ${collec}
                LET record = DOCUMENT(
                    v.${literal(NODE_LIBRARY_ID_FIELD)},
                    v.${literal(NODE_RECORD_ID_FIELD)}
                )
                RETURN record._id
        `;

        return queryPart;
    };
}
