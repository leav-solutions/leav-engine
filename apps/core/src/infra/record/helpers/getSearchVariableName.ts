// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import camelCase from 'lodash/camelCase';
import {IRecordFilterOption} from '../../../_types/record';
import {IFilterTypesHelper} from './filterTypes';

interface IDeps {
    'core.infra.record.helpers.filterTypes'?: IFilterTypesHelper;
}

export type GetSearchVariableName = (filter: IRecordFilterOption) => string;

export default function ({
    'core.infra.record.helpers.filterTypes': filterTypesHelper = null
}: IDeps): GetSearchVariableName {
    const {isAttributeFilter, isClassifyingFilter} = filterTypesHelper;

    return filter => {
        if (isAttributeFilter(filter)) {
            const attributesNames = filter.attributes.map(attribute => camelCase(attribute.id));

            return `${attributesNames.join('_')}_Value`;
        } else if (isClassifyingFilter(filter)) {
            const treeName = camelCase(filter.treeId);

            return `classified_${treeName}_${filter.value}`;
        }
    };
}
