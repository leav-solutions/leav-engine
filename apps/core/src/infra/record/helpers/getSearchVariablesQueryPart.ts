// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {GeneratedAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IAttributeTypesRepo} from 'infra/attributeTypes/attributeTypesRepo';
import {IRecordFilterOption} from '../../../_types/record';
import {IFilterTypesHelper} from './filterTypes';
import {GetClassifyingFiltersVariableQueryPart} from './getClassifyingFiltersVariableQueryPart';
import {GetSearchVariableName} from './getSearchVariableName';

interface IDeps {
    'core.infra.attributeTypes'?: IAttributeTypesRepo;
    'core.infra.record.helpers.getClassifyingFiltersVariableQueryPart'?: GetClassifyingFiltersVariableQueryPart;
    'core.infra.record.helpers.getSearchVariableName'?: GetSearchVariableName;
    'core.infra.record.helpers.filterTypes'?: IFilterTypesHelper;
}

export type GetSearchVariablesQueryPart = (filters: IRecordFilterOption[]) => GeneratedAqlQuery[];

export default function ({
    'core.infra.attributeTypes': attributeTypes = null,
    'core.infra.record.helpers.getClassifyingFiltersVariableQueryPart': getClassifyingFiltersVariableQueryPart = null,
    'core.infra.record.helpers.getSearchVariableName': getSearchVariableName = null,
    'core.infra.record.helpers.filterTypes': filterTypesHelper = null
}: IDeps): GetSearchVariablesQueryPart {
    const {isAttributeFilter, isClassifyingFilter} = filterTypesHelper;

    return filters => {
        const queryPartsById: {[id: string]: GeneratedAqlQuery} = {};

        const variablesDeclarations = filters.reduce((acc, filter): GeneratedAqlQuery[] => {
            if (!isAttributeFilter(filter) && !isClassifyingFilter(filter)) {
                return acc;
            }

            const variableName = getSearchVariableName(filter);
            if (typeof queryPartsById[variableName] !== 'undefined') {
                return acc;
            }

            let variablePart;
            if (isAttributeFilter(filter)) {
                const lastQueryAttribute = filter.attributes[0];
                const typeRepo = attributeTypes.getTypeRepo(lastQueryAttribute);

                variablePart = typeRepo.filterValueQueryPart(
                    filter.attributes.map(attr => ({...attr, _repo: attributeTypes.getTypeRepo(attr)})),
                    filter
                );
            } else if (isClassifyingFilter(filter)) {
                variablePart = getClassifyingFiltersVariableQueryPart(filter);
            }

            queryPartsById[variableName] = variablePart;
            acc.push(aql`LET ${aql.literal(variableName)} = (${variablePart})`);

            return acc;
        }, []);

        return variablesDeclarations;
    };
}
