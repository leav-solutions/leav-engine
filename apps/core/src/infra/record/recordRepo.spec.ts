// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql, Database} from 'arangojs';
import {cloneDeep} from 'lodash';
import {AttributeTypes} from '../../_types/attribute';
import {AttributeCondition, IRecordFilterOption, Operator} from '../../_types/record';
import {IAttributeTypeRepo, IAttributeTypesRepo} from '../attributeTypes/attributeTypesRepo';
import {IDbUtils} from '../db/dbUtils';
import {IFilterTypesHelper} from './helpers/filterTypes';
import recordRepo from './recordRepo';

describe('RecordRepo', () => {
    const ctx = {
        userId: '0',
        requestId: '123465'
    };
    describe('createRecord', () => {
        test('Should create a new record', async function () {
            const recordData = {created_at: 1519303348, modified_at: 1519303348};
            const createdRecordData = {
                _id: 'users/222435651',
                _rev: '_WSywvyC--_',
                _key: 222435651,
                created_at: 1519303348,
                modified_at: 1519303348
            };

            const cleanCreatedRecordData = {
                id: '222435651',
                library: 'users',
                created_at: 1519303348,
                modified_at: 1519303348
            };

            const mockDbCollec = {
                save: global.__mockPromise(createdRecordData),
                document: global.__mockPromise(createdRecordData)
            };

            const mockDb = new Database();
            mockDb.collection = jest.fn().mockReturnValue(mockDbCollec);

            const mockDbServ = {db: mockDb};

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(cleanCreatedRecordData)
            };

            const recRepo = recordRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const createdRecord = await recRepo.createRecord({libraryId: 'test', recordData, ctx});
            expect(mockDbCollec.save.mock.calls.length).toBe(1);
            expect(mockDbCollec.save).toBeCalledWith(recordData);

            expect(mockDbUtils.cleanup.mock.calls.length).toBe(1);
            expect(mockDbUtils.cleanup.mock.calls[0][0].hasOwnProperty('library')).toBe(true);

            expect(createdRecord).toMatchObject(cleanCreatedRecordData);
        });
    });

    describe('updateRecord', () => {
        test('Should update a record', async function () {
            const recordData = {id: '222435651', modified_at: 1519303348};
            const updatedRecordData = {
                _id: 'users/222435651',
                _rev: '_WSywvyC--_',
                _key: 222435651,
                created_at: 1519303348,
                modified_at: 1519303348
            };

            const cleanUpdatedRecordData = {
                id: '222435651',
                library: 'users',
                created_at: 1519303348,
                modified_at: 1519303348
            };

            const mockDbCollec = {
                update: global.__mockPromise(updatedRecordData)
            };

            const mockDbServ = {db: new Database()};
            mockDbServ.db.collection = jest.fn().mockReturnValue(mockDbCollec);

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(cleanUpdatedRecordData)
            };

            const recRepo = recordRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const updatedRecord = await recRepo.updateRecord({libraryId: 'test', recordData, ctx});

            expect(mockDbCollec.update.mock.calls.length).toBe(1);
            expect(mockDbCollec.update).toBeCalledWith(
                {_key: String(recordData.id)},
                {...recordData, id: undefined},
                {returnOld: true}
            );

            expect(mockDbUtils.cleanup.mock.calls.length).toBe(1);

            expect(updatedRecord).toMatchObject(cleanUpdatedRecordData);
        });
    });

    describe('deleteRecord', () => {
        test('Should delete a record and return deleted record', async function () {
            const recordData = {id: '222435651', created_at: 1519303348, modified_at: 1519303348};
            const deletedRecordData = {
                _id: 'users/222435651',
                _rev: '_WSywvyC--_',
                _key: 222435651,
                created_at: 1519303348,
                modified_at: 1519303348
            };

            const mockDbCollec = {
                remove: global.__mockPromise(deletedRecordData)
            };

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([
                    {
                        _key: '222612340',
                        _id: 'core_edge_values_links/222612340',
                        _from: 'ubs/222536515',
                        _to: 'core_values/222612335',
                        _rev: '_Wf3oPWC--_',
                        attribute: 'label',
                        modified_at: 1521047926,
                        created_at: 1521047926
                    },
                    {
                        _key: '223188816',
                        _id: 'core_edge_values_links/223188816',
                        _from: 'products/222763208',
                        _to: 'ubs/222536515',
                        _rev: '_WlGSULm--_',
                        attribute: 'linkedUb',
                        modified_at: 1522936384,
                        created_at: 1522936384
                    }
                ])
            };
            mockDbServ.db.collection = jest.fn().mockReturnValue(mockDbCollec);

            const mockDbUtils: Mockify<IDbUtils> = {cleanup: jest.fn().mockReturnValue(recordData)};

            const recRepo = recordRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const deleteRes = await recRepo.deleteRecord({libraryId: 'users', recordId: recordData.id, ctx});

            expect(mockDbCollec.remove.mock.calls.length).toBe(1);
            expect(mockDbCollec.remove).toBeCalledWith({_key: String(recordData.id)}, {returnOld: true});

            expect(mockDbUtils.cleanup.mock.calls.length).toBe(2);

            expect(deleteRes).toMatchObject(recordData);
        });
    });

    describe('find', () => {
        test('Should find records', async function () {
            const mockQueryRes = {
                totalCount: 2,
                results: [
                    {
                        _key: '222536283',
                        _id: 'ubs/222536283',
                        _rev: '_WgM_51a--_',
                        created_at: 1520931427,
                        modified_at: 1520931427,
                        ean: '9876543219999999',
                        visual_simple: '222713677'
                    },
                    {
                        _key: '222536515',
                        _id: 'ubs/222536515',
                        _rev: '_WgFARB6--_',
                        created_at: 1520931648,
                        modified_at: 1520931648,
                        ean: '9876543219999999'
                    }
                ]
            };

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(mockQueryRes)
            };

            const mockCleanupRes = [
                {
                    id: '222536283',
                    created_at: 1520931427,
                    modified_at: 1520931427,
                    ean: '9876543219999999',
                    visual_simple: '222713677'
                },
                {
                    id: '222536515',
                    created_at: 1520931648,
                    modified_at: 1520931648,
                    ean: '9876543219999999'
                }
            ];

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValueOnce(mockCleanupRes[0]).mockReturnValueOnce(mockCleanupRes[1])
            };

            const recRepo = recordRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const records = await recRepo.find({
                libraryId: 'test_lib',
                filters: [],
                pagination: null,
                withCount: true,
                ctx
            });
            expect(mockDbServ.execute.mock.calls.length).toBe(1);

            expect(mockDbServ.execute.mock.calls[0][0]).toMatchSnapshot();

            expect(records).toEqual({
                cursor: null,
                totalCount: 2,
                list: mockCleanupRes
            });
        });

        test('Should paginate with offset', async function () {
            const mockQueryRes = {
                totalCount: 5,
                results: [
                    {
                        _key: '222536283',
                        _id: 'ubs/222536283',
                        _rev: '_WgM_51a--_',
                        created_at: 1520931427,
                        modified_at: 1520931427,
                        ean: '9876543219999999',
                        visual_simple: '222713677'
                    },
                    {
                        _key: '222536515',
                        _id: 'ubs/222536515',
                        _rev: '_WgFARB6--_',
                        created_at: 1520931648,
                        modified_at: 1520931648,
                        ean: '9876543219999999'
                    }
                ]
            };

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(mockQueryRes)
            };

            const mockCleanupRes = [
                {
                    id: '222536283',
                    created_at: 1520931427,
                    modified_at: 1520931427,
                    ean: '9876543219999999',
                    visual_simple: '222713677'
                },
                {
                    id: '222536515',
                    created_at: 1520931648,
                    modified_at: 1520931648,
                    ean: '9876543219999999'
                }
            ];

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValueOnce(mockCleanupRes[0]).mockReturnValueOnce(mockCleanupRes[1])
            };

            const recRepo = recordRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const records = await recRepo.find({
                libraryId: 'test_lib',
                filters: [],
                pagination: {limit: 2, offset: 0},
                withCount: true,
                ctx
            });
            expect(mockDbServ.execute.mock.calls.length).toBe(1);

            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch('LIMIT');

            expect(records.totalCount).toBe(5);
            expect(records.list.length).toBe(2);
        });

        test('Should paginate with cursor', async function () {
            const mockQueryRes = [
                {
                    _key: '222536283',
                    _id: 'ubs/222536283',
                    _rev: '_WgM_51a--_',
                    created_at: 1520931427,
                    modified_at: 1520931427,
                    ean: '9876543219999999',
                    visual_simple: '222713677'
                },
                {
                    _key: '222536515',
                    _id: 'ubs/222536515',
                    _rev: '_WgFARB6--_',
                    created_at: 1520931648,
                    modified_at: 1520931648,
                    ean: '9876543219999999'
                }
            ];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(mockQueryRes)
            };

            const mockCleanupRes = [
                {
                    id: '222536283',
                    created_at: 1520931427,
                    modified_at: 1520931427,
                    ean: '9876543219999999',
                    visual_simple: '222713677'
                },
                {
                    id: '222536515',
                    created_at: 1520931648,
                    modified_at: 1520931648,
                    ean: '9876543219999999'
                }
            ];

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValueOnce(mockCleanupRes[0]).mockReturnValueOnce(mockCleanupRes[1])
            };

            const recRepo = recordRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const records = await recRepo.find({
                libraryId: 'test_lib',
                filters: [],
                pagination: {limit: 2, cursor: 'bmV4dDoyOjEzNDYzNDQ0'},
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch('LIMIT');
            expect(mockDbServ.execute.mock.calls[0][0].withTotalCount).not.toBe(true); // No count

            expect(records.list.length).toBe(2);
            expect(records.cursor.next).toBeTruthy();
        });

        describe('Inactive records', () => {
            const mockQueryRes = [
                {
                    _key: '222536283',
                    _id: 'ubs/222536283',
                    _rev: '_WgM_51a--_',
                    created_at: 1520931427,
                    modified_at: 1520931427
                }
            ];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(mockQueryRes)
            };

            const mockCleanupRes = {
                id: '222536283',
                created_at: 1520931427,
                modified_at: 1520931427
            };

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(mockCleanupRes)
            };

            const recRepo = recordRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            beforeEach(() => jest.clearAllMocks());

            test('Should not retrieve inactive records', async () => {
                const records = await recRepo.find({
                    libraryId: 'test_lib',
                    filters: [],
                    pagination: null,
                    withCount: false,
                    retrieveInactive: false,
                    ctx
                });
                expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch('active == true');
            });

            test('Should retrieve inactive records if forced', async () => {
                const records = await recRepo.find({
                    libraryId: 'test_lib',
                    filters: [],
                    pagination: null,
                    withCount: false,
                    retrieveInactive: true,
                    ctx
                });
                expect(mockDbServ.execute.mock.calls[0][0].query.query).not.toMatch('active == true');
            });
        });
    });

    describe('find with filters', () => {
        const mockFilters: IRecordFilterOption[] = [
            {
                attributes: [
                    {
                        id: 'test_attr',
                        type: null
                    }
                ],
                condition: AttributeCondition.EQUAL,
                value: 'test'
            },
            {operator: Operator.AND},
            {
                attributes: [
                    {
                        id: 'test_attr2',
                        type: null
                    }
                ],
                condition: AttributeCondition.EQUAL,
                value: 'test2'
            }
        ];

        test('Should filter records - simple', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise({
                    totalCount: 1,
                    results: [
                        {
                            _key: '222536515',
                            _id: 'test_lib/222536515',
                            _rev: '_WgM_51a--_',
                            created_at: 1520931427,
                            modified_at: 1520931427,
                            test_attr: 'test'
                        }
                    ]
                })
            };

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue({
                    id: '222536515',
                    created_at: 1520931427,
                    modified_at: 1520931427,
                    test_attr: 'test'
                })
            };

            const mockAttrSimpleRepo: Mockify<IAttributeTypeRepo> = {};

            const mockAttrRepo: Mockify<IAttributeTypesRepo> = {
                getTypeRepo: jest.fn().mockReturnValue(mockAttrSimpleRepo as IAttributeTypesRepo)
            };

            const mockGetSearchVariablesQueryPart = jest.fn().mockReturnValue(aql`<Variables>`);
            const mockGetSearchVariableName = jest.fn().mockReturnValue(aql`variableName`);
            const mockGetConditionPart = jest.fn().mockReturnValue(aql`<Condition>`);

            const mockFilterTypesHelper: Mockify<IFilterTypesHelper> = {
                isCountFilter: jest.fn().mockReturnValue(false),
                isAttributeFilter: jest.fn().mockReturnValue(true)
            };

            const recRepo = recordRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils,
                'core.infra.attributeTypes': mockAttrRepo as IAttributeTypesRepo,
                'core.infra.record.helpers.getSearchVariablesQueryPart': mockGetSearchVariablesQueryPart,
                'core.infra.record.helpers.getSearchVariableName': mockGetSearchVariableName,
                'core.infra.attributeTypes.helpers.getConditionPart': mockGetConditionPart,
                'core.infra.record.helpers.filterTypes': mockFilterTypesHelper as IFilterTypesHelper
            });

            const filters = cloneDeep(mockFilters);
            filters[0].attributes[0].type = AttributeTypes.SIMPLE;
            filters[2].attributes[0].type = AttributeTypes.SIMPLE;

            const records = await recRepo.find({
                libraryId: 'test_lib',
                filters,
                pagination: null,
                withCount: true,
                ctx
            });

            expect(mockDbServ.execute.mock.calls[0][0]).toMatchSnapshot();
            expect(records).toEqual({
                cursor: null,
                totalCount: 1,
                list: [
                    {
                        id: '222536515',
                        created_at: 1520931427,
                        modified_at: 1520931427,
                        test_attr: 'test'
                    }
                ]
            });
        });
    });

    describe('search', () => {
        test('should search records', async function () {
            const mockQueryRes = {
                took: 322,
                timed_out: false,
                _shards: {total: 1, successful: 1, skipped: 0, failed: 0},
                hits: {
                    total: {value: 1, relation: 'eq'},
                    max_score: 0.2876821,
                    hits: [
                        {
                            _id: 1,
                            _source: {
                                library: 'test_lib'
                            }
                        }
                    ]
                }
            };

            const mockElasticsearch = {
                wildcardSearch: global.__mockPromise(mockQueryRes)
            };

            const recRepo = recordRepo({
                'core.infra.elasticsearch.elasticsearchService': mockElasticsearch
            });

            const result = await recRepo.search('test_lib', 'text');

            expect(mockElasticsearch.wildcardSearch.mock.calls.length).toBe(1);
            expect(mockElasticsearch.wildcardSearch.mock.calls[0][0]).toMatchSnapshot();

            expect(result).toEqual({
                totalCount: 1,
                list: [
                    {
                        id: 1,
                        library: 'test_lib'
                    }
                ]
            });
        });

        test('should search record with from/size params', async function () {
            const mockQueryRes = {
                took: 322,
                timed_out: false,
                _shards: {total: 1, successful: 1, skipped: 0, failed: 0},
                hits: {
                    total: {value: 2, relation: 'eq'},
                    max_score: 0.2876821,
                    hits: [
                        {
                            _id: 1,
                            _source: {
                                library: 'test_lib'
                            }
                        }
                    ]
                }
            };

            const mockElasticsearch = {
                wildcardSearch: global.__mockPromise(mockQueryRes)
            };

            const recRepo = recordRepo({
                'core.infra.elasticsearch.elasticsearchService': mockElasticsearch
            });

            const result = await recRepo.search('test_lib', 'text', 0, 1);

            expect(mockElasticsearch.wildcardSearch.mock.calls.length).toBe(1);
            expect(mockElasticsearch.wildcardSearch.mock.calls[0][0]).toMatchSnapshot();

            expect(result).toEqual({
                totalCount: 2,
                list: [
                    {
                        id: 1,
                        library: 'test_lib'
                    }
                ]
            });
        });
    });
});
