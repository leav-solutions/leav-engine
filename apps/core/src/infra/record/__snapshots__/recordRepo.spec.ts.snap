// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordRepo find Handle fulltext search result 1`] = `
{
  "ctx": {
    "requestId": "123465",
    "userId": "0",
  },
  "query": {
    "_source": [Function],
    "bindVars": {
      "@value0": "test_lib",
      "value1": [
        "123456",
        "789012",
      ],
      "value2": "test_lib",
    },
    "query": "FOR r IN @@value0
FILTER r._key IN @value1
SORT r._key ASC
FILTER r.active == true
RETURN MERGE(r, {library: @value2})",
  },
  "withTotalCount": true,
}
`;

exports[`RecordRepo find Should find records 1`] = `
{
  "ctx": {
    "requestId": "123465",
    "userId": "0",
  },
  "query": {
    "_source": [Function],
    "bindVars": {
      "@value0": "test_lib",
      "value1": "test_lib",
    },
    "query": "FOR r IN @@value0
SORT r._key ASC
FILTER r.active == true
RETURN MERGE(r, {library: @value1})",
  },
  "withTotalCount": true,
}
`;

exports[`RecordRepo find with filters Should filter records - simple 1`] = `
{
  "ctx": {
    "requestId": "123465",
    "userId": "0",
  },
  "query": {
    "_source": [Function],
    "bindVars": {
      "@value0": "test_lib",
      "value1": "test_lib",
    },
    "query": "FOR r IN @@value0

FILTER (
( IS_ARRAY([object Object]) ? LENGTH([object Object][* FILTER <Condition>]) : <Condition> )
AND
( IS_ARRAY([object Object]) ? LENGTH([object Object][* FILTER <Condition>]) : <Condition> )
)
SORT r._key ASC
FILTER r.active == true
RETURN MERGE(r, {library: @value1})",
  },
  "withTotalCount": true,
}
`;

exports[`RecordRepo search should search record with from/size params 1`] = `"test_lib"`;

exports[`RecordRepo search should search records 1`] = `"test_lib"`;
